File: server/hotrod/src/test/scala/org/infinispan/server/hotrod/HotRodConcurrentStartTest.scala
package org.infinispan.server.hotrod

import org.infinispan.configuration.cache.CacheMode
import org.infinispan.test.MultipleCacheManagersTest
import org.infinispan.test.fwk.TestCacheManagerFactory
import org.testng.annotations.Test

import scala.concurrent.duration._
import scala.concurrent.{Await, ExecutionContext, Future}
import scala.util.{Failure, Success, Try}

// Do not remove, otherwise getDefaultClusteredConfig is not found
import org.infinispan.server.core.test.ServerTestingUtil._
import org.infinispan.server.hotrod.test.HotRodTestingUtil._
import org.infinispan.server.hotrod.test.UniquePortThreadLocal
import org.infinispan.test.AbstractCacheTest._

/**
 * Tests concurrent Hot Rod server startups
 *
 * @author Galder Zamarreño
 * @since 4.2
 */
@Test(groups = Array("functional"), testName = "server.hotrod.HotRodConcurrentStartTest")
class HotRodConcurrentStartTest extends MultipleCacheManagersTest {
   private val numberOfServers = 2
   private val cacheName = "hotRodConcurrentStart"

   @Test(enabled=false) // Disable explicitly to avoid TestNG thinking this is a test!!
   override def createCacheManagers() {
      for (i <- 0 until numberOfServers) {
         val cm = TestCacheManagerFactory.createClusteredCacheManager(hotRodCacheConfiguration())
         cacheManagers.add(cm)
         val cfg = hotRodCacheConfiguration(
            getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, false))
         cm.defineConfiguration(cacheName, cfg.build())
      }
   }

   def testConcurrentStartup() {
      val initialPort = UniquePortThreadLocal.get.intValue

      implicit val forExecutionContext = new ExecutionContext {
         override def execute(runnable: Runnable): Unit = fork(runnable)
         override def reportFailure(cause: Throwable): Unit = throw cause
      }

      val futures: Seq[Future[HotRodServer]] = (1 to numberOfServers).map {
         case 1 =>  Future(startHotRodServerWithDelay(getCacheManagers.get(0), initialPort, 10000))
         case i =>  Future(startHotRodServer(getCacheManagers.get(i - 1), initialPort + (i * 10)))
      }

      val results = futures.map(f => Try(Await.result(f, 1 minute)))
      val success = results.collect { case Success(s) => s}
      val errors = results.collect { case Failure(e) => e}
      if (success.isEmpty) {
         errors.foreach(log.error("Server failed to start", _))
         throw new AssertionError("All serves failed to start, see error log messages for more info")
      }
      // kill servers that started
      success map killServer
      // throw error if any
      errors map(throw _)
   }

}

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/HotRodConcurrentStartTest.java
package org.infinispan.server.hotrod;

import static org.infinispan.server.hotrod.test.HotRodTestingUtil.hotRodCacheConfiguration;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.serverPort;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.startHotRodServer;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import org.infinispan.commands.module.TestGlobalConfigurationBuilder;
import org.infinispan.configuration.cache.CacheMode;
import org.infinispan.configuration.cache.ConfigurationBuilder;
import org.infinispan.configuration.global.GlobalConfigurationBuilder;
import org.infinispan.manager.EmbeddedCacheManager;
import org.infinispan.remoting.transport.Transport;
import org.infinispan.server.core.test.ServerTestingUtil;
import org.infinispan.server.hotrod.configuration.HotRodServerConfiguration;
import org.infinispan.server.hotrod.test.HotRodClient;
import org.infinispan.test.MultipleCacheManagersTest;
import org.infinispan.test.TestingUtil;
import org.infinispan.test.fwk.TestCacheManagerFactory;
import org.testng.annotations.Test;

/**
 * Tests concurrent Hot Rod server startups
 *
 * @author Galder Zamarreño
 * @since 4.2
 */
@Test(groups = "functional", testName = "server.hotrod.HotRodConcurrentStartTest")
public class HotRodConcurrentStartTest extends MultipleCacheManagersTest {
   public static final int NUMBER_OF_SERVERS = 2;
   public static final String CACHE_NAME = "hotRodConcurrentStart";

   @Override
   protected void createCacheManagers() {
      for (int i = 0; i < NUMBER_OF_SERVERS; i++) {
         GlobalConfigurationBuilder globalConfig = GlobalConfigurationBuilder.defaultClusteredBuilder();
         if (i == 0) {
            // Delay the start of the first server's address cache
            globalConfig.addModule(TestGlobalConfigurationBuilder.class).cacheStartingCallback(cr -> {
               if (cr.getCacheName().startsWith(HotRodServerConfiguration.TOPOLOGY_CACHE_NAME_PREFIX)) {
                  log.tracef("Delaying start of cache %s on %s", cr.getCacheName(),
                             cr.getComponent(Transport.class).getAddress());
                  TestingUtil.sleepThread(1000);
               }
            });
         }
         EmbeddedCacheManager cm = TestCacheManagerFactory.createClusteredCacheManager(globalConfig, null);
         cacheManagers.add(cm);
         ConfigurationBuilder cfg =
               hotRodCacheConfiguration(getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, false));
         cm.defineConfiguration(CACHE_NAME, cfg.build());
      }
   }

   public void testConcurrentStartup(Method m) throws Exception {
      int initialPort = serverPort();

      List<HotRodServer> servers = new ArrayList<>();
      try {
         List<Future<HotRodServer>> futures = new ArrayList<>();
         for (int i = 0; i < NUMBER_OF_SERVERS; i++) {
            int finalI = i;
            futures.add(fork(() -> {
               HotRodServer server = startHotRodServer(getCacheManagers().get(finalI), initialPort + (finalI * 10));
               servers.add(server);
               return server;
            }));
         }

         for (Future<HotRodServer> hotRodServerFuture : futures) {
            hotRodServerFuture.get(30, TimeUnit.SECONDS);
         }

         try (HotRodClient client = new HotRodClient(servers.get(0).getHost(), servers.get(0).getPort(),
                                                     CACHE_NAME, HotRodConstants.VERSION_30)) {
            client.assertPut(m);
         }
      } finally {
         servers.forEach(ServerTestingUtil::killServer);
      }
   }

}
